src/storage/src/table/cell_based_table.rs

```diff
@@ -359,45 +420,97 @@ impl<S: PkAndRowStream + Unpin> TableIter for S {
 }
 
 /// Iterators
-impl<S: StateStore> CellBasedTable<S> {
+impl<S: StateStore, const T: AccessType> CellBasedTable<S, T> {
+    /// Get multiple [`CellBasedIter`] based on the specified vnodes.
+    async fn iters_with_encoded_key_range<R, B, const ITER_TYPE: bool>(
+        &self,
+        encoded_key_range: R,
+        epoch: u64,
+    ) -> StorageResult<Vec<CellBasedIter<S, ITER_TYPE>>>
+    where
+        R: RangeBounds<B> + Send + Clone,
+        B: AsRef<[u8]> + Send,
+    {
+        // For each vnode, construct an iterator.
+        // TODO: if there're some vnodes continuously in the range and we don't care about order, we
+        // can use a single iterator.
+        let mut iterators = Vec::with_capacity(self.vnodes.num_high_bits());
+
+        for vnode in self
+            .vnodes
+            .iter()
+            .enumerate()
+            .filter(|&(_, set)| set)
+            .map(|(i, _)| i as VirtualNode)
+        {
+            let raw_key_range = prefixed_range(encoded_key_range.clone(), &vnode.to_be_bytes());
+            let iter = CellBasedIter::<_, ITER_TYPE>::new(
+                &self.keyspace,
+                self.mapping.clone(),
+                raw_key_range,
+                epoch,
+            )
+            .await?;
+            iterators.push(iter);
+        }
+
+        Ok(iterators)
+    }
+
     /// Get a [`StreamingIter`] with given `encoded_key_range`.
     pub(super) async fn streaming_iter_with_encoded_key_range<R, B>(
         &self,
         encoded_key_range: R,
         epoch: u64,
     ) -> StorageResult<StreamingIter<S>>
     where
-        R: RangeBounds<B> + Send,
+        R: RangeBounds<B> + Send + Clone,
         B: AsRef<[u8]> + Send,
     {
-        Ok(CellBasedIter::<_, STREAMING_ITER_TYPE>::new(
-            &self.keyspace,
-            self.mapping.clone(),
-            encoded_key_range,
-            epoch,
-        )
-        .await?
-        .into_stream())
+        let iterators = self
+            .iters_with_encoded_key_range::<_, _, STREAMING_ITER_TYPE>(encoded_key_range, epoch)
+            .await?
+            .into_iter()
+            .map(|i| i.into_stream())
+            .collect_vec();
+
+        let iter = match iterators.len() {
+            0 => unreachable!(),
+            1 => iterators.into_iter().next().unwrap(),
+            _ => todo!("merge multiple vnode ranges"),
```

Comment:
What's the relationship between `iterators.len()` and vonde nums? Is it one-on-one?

Reply:
It depends on whether we need to preserve the order of primary keys. If so, we need to construct an iterator for every single vnode and merge them into a sorted one, or we can construct a single for a continuous range.